controllers/clustertopology_controller_sync_test.go

```diff
@@ -0,0 +1,294 @@
+		},
+
+		// Field only in modified (not existing in original) --> align to modified
+
+		{
+			name: "Field only in modified, align to modified",
+			original: &unstructured.Unstructured{ // current
+				Object: map[string]interface{}{},
+			},
+			modified: &unstructured.Unstructured{ // desired
+				Object: map[string]interface{}{
+					"A": "A",
+				},
+			},
+			wantHasChanges: true,
+			wantPatch:      []byte("{\"A\":\"A\"}"),
+		},
+		{
+			name: "Nested field only in modified, align to modified",
+			original: &unstructured.Unstructured{ // current
+				Object: map[string]interface{}{},
+			},
+			modified: &unstructured.Unstructured{ // desired
+				Object: map[string]interface{}{
+					"spec": map[string]interface{}{
+						"template": map[string]interface{}{
+							"spec": map[string]interface{}{
+								"A": "A",
+							},
+						},
+					},
+				},
+			},
+			wantHasChanges: true,
+			wantPatch:      []byte("{\"spec\":{\"template\":{\"spec\":{\"A\":\"A\"}}}}"),
+		},
+
+		// Field only in original (not existing in modified) --> preserve original
+
+		{
+			name: "Field only in modified, align to modified",
+			original: &unstructured.Unstructured{ // current
+				Object: map[string]interface{}{
+					"A": "A",
+				},
+			},
+			modified: &unstructured.Unstructured{ // desired
+				Object: map[string]interface{}{},
+			},
+			wantHasChanges: false,
+			wantPatch:      []byte("{}"),
+		},
+		{
+			name: "Nested field only in modified, align to modified",
+			original: &unstructured.Unstructured{ // current
+				Object: map[string]interface{}{
+					"spec": map[string]interface{}{
+						"template": map[string]interface{}{
+							"spec": map[string]interface{}{
+								"A": "A",
+							},
+						},
+					},
+				},
+			},
+			modified: &unstructured.Unstructured{ // desired
+				Object: map[string]interface{}{},
+			},
+			wantHasChanges: false,
+			wantPatch:      []byte("{}"),
+		},
+
+		// More tests
+		{
+			name: "NoChanges",
+			original: &unstructured.Unstructured{
+				Object: map[string]interface{}{
+					"spec": map[string]interface{}{
+						"A": "A",
+						"B": "B",
+						"C": "C", // C only in modified
+					},
+				},
+			},
+			modified: &unstructured.Unstructured{
+				Object: map[string]interface{}{
+					"spec": map[string]interface{}{
+						"A": "A",
+						"B": "B",
+					},
+				},
+			},
+			wantHasChanges: false,
+			wantPatch:      []byte("{}"),
+		},
+		{
+			name: "Many changes",
+			original: &unstructured.Unstructured{
+				Object: map[string]interface{}{
+					"spec": map[string]interface{}{
+						"A": "A",
+						// B missing
+						"C": "C", // C only in modified
+					},
+				},
+			},
+			modified: &unstructured.Unstructured{
+				Object: map[string]interface{}{
+					"spec": map[string]interface{}{
+						"A": "A",
+						"B": "B",
+					},
+				},
+			},
+			wantHasChanges: true,
+			wantPatch:      []byte("{\"spec\":{\"B\":\"B\"}}"),
+		},
```

Comment:
Would it make sense to add a few tests where we use real structs? This would allow us to also test how pointer affect the patches.

Reply:
At the moment we planning to use this func only with unstructured types; I would defer adding typed test to the future if the above assumption changes.
Also, I don't think we should test how json serialisation works, and things like pointers or omitempty all boils up to the field existing or not, which is what we are testing.